subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,166 @@
+}
+
+// Stop stops advertising a bluetooth service which (when enabled) accepts WiFi and Viam cloud config credentials.
+func (bwp *bluetoothService) stop() error { //nolint:unused
+	return errors.New("TODO APP-7651: Implement helper methods to start/stop advertising BLE connection")
+}
+
+// Update updates the list of networks that are advertised via bluetooth as available.
+func (bwp *bluetoothService) refreshAvailableNetworks(ctx context.Context, awns []*NetworkInfo) error { //nolint:unused
+	return errors.New("TODO APP-7652: Implement helper method to write update WiFi networks to BLE peripheral characteristic")
+}
+
+// WaitForCredentials returns credentials, the minimum required information to provision a robot and/or its WiFi.
+func (bwp *bluetoothService) waitForCredentials( //nolint:unused
+	ctx context.Context, requiresCloudCredentials, requiresWiFiCredentials bool,
+) (*userInput, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if !requiresWiFiCredentials && !requiresCloudCredentials {
+		return nil, errors.New("should be waiting for cloud credentials or WiFi credentials, or both")
+	}
+	var ssid, psk, robotPartKeyID, robotPartKey string
+	var ctxErr, ssidErr, pskErr, robotPartKeyIDErr, robotPartKeyErr error
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	utils.ManagedGo(func() {
+		for {
+			if ctxErr = ctx.Err(); ctxErr != nil {
+				return
+			}
+			select {
+			case <-ctx.Done():
+				ctxErr = ctx.Err()
+				return
+			default:
+				if requiresWiFiCredentials {
+					if ssid == "" {
+						var e *emptyBluetoothCharacteristicError
+						ssid, ssidErr = bwp.readSsid()
+						if ssidErr != nil && !errors.As(ssidErr, &e) {
+							return
+						}
+					}
+					if psk == "" {
+						var e *emptyBluetoothCharacteristicError
+						psk, pskErr = bwp.readPsk()
+						if pskErr != nil && !errors.As(pskErr, &e) {
+							return
+						}
+					}
+				}
+				if requiresCloudCredentials {
+					if robotPartKeyID == "" {
+						var e *emptyBluetoothCharacteristicError
+						robotPartKeyID, robotPartKeyIDErr = bwp.readRobotPartKeyID()
+						if robotPartKeyIDErr != nil && !errors.As(robotPartKeyIDErr, &e) {
+							return
+						}
+					}
+					if robotPartKey == "" {
+						var e *emptyBluetoothCharacteristicError
+						robotPartKey, robotPartKeyErr = bwp.readRobotPartKey()
+						if robotPartKeyErr != nil && !errors.As(robotPartKeyErr, &e) {
+							return
+						}
+					}
+				}
+				if requiresWiFiCredentials && requiresCloudCredentials && //nolint:gocritic
+					ssid != "" && psk != "" && robotPartKeyID != "" && robotPartKey != "" {
+					return
+				} else if requiresWiFiCredentials && ssid != "" && psk != "" {
+					return
+				} else if requiresCloudCredentials && robotPartKeyID != "" && robotPartKey != "" {
+					return
+				}
+
+				// Not ready to return (do not have the minimum required set of credentials), so sleep and try again.
+				time.Sleep(time.Second)
```

Comment:
Don't use time.Sleep in agent/subsystems, it can cause health checks to fail. If you're inside a loop, use the mainLoopHealth/bgLoopHealth.Sleep(ctx) function instead